### PR TITLE
Fix multiline and add symbolic strings

### PR DESCRIPTION
**Pull request by vkleen**
This PR adjust the multiline string syntax to align with the current syntax in Nickel. On the way it also adds support for the new tagged symbolic strings.

### CODE REVIEW

**Comment by vkleen** [comment]
Actually, don't try to merge this yet. The following
```
tree-sitter parse <(<<<'nix-s% %"abc"%')
```
produces a segfault.

**Comment by vkleen** [comment]
The new version reworks the multistring start scanner to (hopefully correctly) recognize symbolic string starts. It uses a bespoke finite state automaton to recognize a symbolic string start token which probably warrants careful review.

But at least it doesn't seem to segfault or produce weird parse trees on ill-formed input.

File: src/scanner.cc [hunk 1]
```diff
@@ -75,14 +77,87 @@ struct Scanner {
     }
   }
 
-  // Scans the multistring start. Assumes that the m indicating a multistring
-  // was already consumed.
+  bool valid_symtag_start(int32_t lookahead) {
```
Comment (yannham) [inline src/scanner.cc:80]: Might deserve a one-liner comment, I think "symtag" could not be obvious in a distant future.

File: src/scanner.cc [hunk 2]
```diff
@@ -75,14 +77,87 @@ struct Scanner {
...
+    return ('a' <= lookahead && lookahead <= 'z') || ('A' <= lookahead && lookahead <= 'Z');
+  }
+
+  bool valid_symtag_middle(int32_t lookahead) {
+    return valid_symtag_start(lookahead) || ('0' <= lookahead && lookahead <= '9') || lookahead == '-' || lookahead == '\'' || lookahead == '_';
+  }
+
+  bool scan_until_sstr_start_end(TSLexer *lexer, bool m_scanned) {
+    enum RecognizerState {
```
Comment (yannham) [inline src/scanner.cc:89]: Same, having a few comment lines to explain why we need that for could be more posterity-proof.

File: src/scanner.cc [hunk 3]
```diff
@@ -94,17 +169,13 @@ struct Scanner {
     }
 
     expected_percent_count.push_back(count);
-
-    // A START is fully scanned when we started with an 'm' (precondition of
```
Comment (yannham) [inline src/scanner.cc:98]: Could we keep the information and in particular precondition and postcondition, although I imagine that they need to be updated?

**Review by yannham — APPROVED** [review]
Beside documentation, LGTM!

File: src/scanner.cc (hunk 3 above)
Comment (vkleen) [reply src/scanner.cc:98]: I've added a little bit of explanation into the function. With this change there is actually no precondition for calling `scan_multstr_start` anymore.

**Pull request merged**